socs/smurf/MultiSlotController.py

```diff
@@ -0,0 +1,118 @@
+from ocs.ocs_client import OCSClient
+
+
+class MultiSlotController:
+    """
+    Simple client controller useful for managing all pysmurf-controller
+    agents that exist on a network. This can dispatch operations and check
+    operation status for any or all pysmurf-controllers on a network.
+
+    Args
+    -----
+    client_args: list
+        List of command line arguments to pass onto the OCSClient objects.
+
+    Attributes
+    ---------------
+    smurfs : dict
+        Dictionary containing OCSClient objects for each pysmurf-controller
+        on the network. This assumes pysmurf controller agents have
+        instance-id's of the form ``pysmurf-controller-<id>`` where id
+        can be the uxm-id or something like ``crate1slot2``. The keys of this
+        dict will be ``<id>``. This is populated from the registry agent on
+        init or using the funciton ``get_active_smurfs``.
+    """
+
+    def __init__(self, client_args=[]):
+        self.client_args = client_args
+        self.reg = OCSClient('registry', args=client_args)
+        self.smurfs = {}
+        self.get_active_smurfs()
+
+    def get_active_smurfs(self):
+        """
+        Obtains a list of active pysmurf-controller agents on a network from
+        the registry agent. Uses this to populate the ``smurfs`` dictionary,
+        which contains an OCSClient object for each.
+        """
+        reg_agents = self.reg.main.status().session['data']
+        self.smurfs = {}
+        for k, v in reg_agents.items():
+            if 'pysmurf-controller' not in k:
+                continue
+            if v['expired']:
+                continue
+
+            instance_id = k.split('.')[-1]
+            smurf_key = instance_id.split('-')[-1]
+            self.smurfs[smurf_key] = OCSClient(instance_id,
+                                               args=self.client_args)
+
+    def start(self, op_name, ids=None, **kwargs):
+        """
+        Starts an operation on any or all pysmurf controllers.
+
+        Args
+        ----
+        op_name : str
+            Name of the operation to start
+        ids : list, optional
+            List of pysmurf-controller id's. This defaults to running on all
+            controllers.
+        **kwargs
+            Any additional keyword arguments are passed onto the start
+            operation as params
+        """
+        if ids is None:
+            ids = list(self.smurfs.keys())
+        elif isinstance(ids, str):
+            ids = [ids]
```

Comment:
Is this meant to catch times a user submits just a single id?

I could imagine something like `.start('uxm_setup', ids='pysmurf-controller-crate1slot1, pysmurf-controller-crate1slot2')` being submitted, which won't work here.

Reply:
Eh it's less about catching mistakes and more that if I'm passing a single ID I find it more convenient to pass directly instead of in a list, though I admit it doesn't really matter. I'm just used to using `np.atleast_1d` in a lot of my sodetlib functions which handles this kind of input. I can remove it if you want though.